Add support for MS Azure Vision OCR

ml5ah opened the pull request:
This PR implements support for MS Azure Vision OCR. This is part of the roadmap for Tarsier and provides an alternative to the existing Google Cloud Vision OCR option. 

It has been reported that MS Azure OCR performs better on webpages and therefore it makes this addition a worthwhile enhancement. 

Please share feedback or potential improvements. 

Issue: https://github.com/reworkd/tarsier/issues/84

asim-shrestha commented:
Very cool. Ideally we'd have some docs or notes for how to get the azure credential

asim-shrestha commented:
also, is @anishhiranandani the same account as you @ml5ah or are you collaborating on this PR?

ml5ah commented:
Sure, @asim-shrestha, I updated the readme file with details of how to create the credentials JSON for Azure as well as a link that shows the UI screen where the credentials are. 

Yes, anishhiranandani and ml5ah are both me. My bad, I think my Mac was logged on to the old account (anishhiranandani).

asim-shrestha commented:
No worries! @ml5ah i'll make an azure account tomorrow and check it out. Thanks for doing this, has been on the back burner for quite some time

asim-shrestha commented:
Would you be interested in doing some additional work in this repo?

awtkns commented on pyproject.toml:
```diff
@@ -13,6 +13,7 @@ python = "^3.11"
 playwright = "^1.39.0"
 selenium = "^4.15.2"
 google-cloud-vision = "^3.4.5"
+azure-ai-vision-imageanalysis = "^1.0.0"
```
ideally we should make these optional dependancies going forward

awtkns commented on tarsier/__main__.py:
```diff
@@ -6,14 +6,21 @@
 
-async def main(credentials_path: str, url: str, verbose: bool) -> None:
+
+async def main(credentials_path: str, ocr_service: str, url: str, verbose: bool) -> None:
```
this will break existing versions, lets make ocr_service and optional dependancy:
```suggestion
async def main(credentials_path: str, url: str, verbose: bool, ocr_service: str = "google") -> None:
```

ml5ah replied:
Thanks for the feedback @awtkns, done.

ml5ah commented:
> Would you be interested in doing some additional work in this repo?

Sure, I can try to spend some time to help out based on my bandwidth. Do you have any specific items in mind?

awtkns commented on tarsier/__main__.py:
```diff
@@ -6,14 +6,21 @@
 
-async def main(credentials_path: str, url: str, verbose: bool) -> None:
+
+async def main(credentials_path: str, url: str, verbose: bool, ocr_service: str = "google") -> None:
```
instead of having ocr_service be type str we should probably make a type:

OcrProvider = Literal["google", "microsoft"]

ml5ah replied:
Sure, did it slightly differently using StrEnum. Defined in tarsier/ocr

awtkns commented on tarsier/__main__.py:
```diff
@@ -6,14 +6,21 @@
         credentials = json.load(f)
 
-    ocr_service = GoogleVisionOCRService(credentials)
+    match ocr_service:
+        case "google":
+            ocr_service = GoogleVisionOCRService(credentials)
+    
+        case "microsoft":
+            ocr_service = MicrosoftAzureOCRService(credentials)
+
```
we need to throw an error here as the default case

ml5ah replied:
Done

awtkns reviewed (commented):
Looking good! Would you be able to share a couple screen shot examples of it in action?

asim-shrestha commented:
@ml5ah I do have some items that are more involved. Here's an open list of them: https://reworkd.notion.site/Tarsier-optimizations-37ad18b8301a4ac085c64adf7533dfbb?pvs=25

How about two phase tagging in tag optimizations?

ml5ah commented:
Here's 2 screenshots of it in action: 

![screenshot](https://github.com/reworkd/tarsier/assets/65744410/170f21ca-7f3c-49e8-b5ed-6e522aa4f0e5)
<img width="1380" alt="Screenshot 2024-06-12 at 15 27 22" src="https://github.com/reworkd/tarsier/assets/65744410/9299af36-4c35-4f8d-ade2-85bf1694e637">

ml5ah commented:
> Looking good! Would you be able to share a couple screen shot examples of it in action?

Sure, @awtkns , shared. I'm having some issues with the poetry.lock file -- not sure why. Are you able to take a quick look? Thanks!

awtkns commented:
Yup, i can fix the poetry lock issues~

awtkns commented:
Thanks for this @ml5ah! Tested, looks good. I removed the strenum dependancy in favor of the built-in Literal type. Also added an e2e test.

ml5ah commented:
> Thanks for this @ml5ah! Tested, looks good. I removed the strenum dependancy in favor of the built-in Literal type. Also added an e2e test.

Great! Thanks @awtkns

ml5ah commented:
> @ml5ah I do have some items that are more involved. Here's an open list of them: https://reworkd.notion.site/Tarsier-optimizations-37ad18b8301a4ac085c64adf7533dfbb?pvs=25
> 
> How about two phase tagging in tag optimizations?

Sure, @asim-shrestha. Sounds like a very interesting problem!

asim-shrestha commented:
Cool @ml5ah! Can you make an issue and we can collaborate / answer questions there
